src/spartan/polys/multilinear.rs

```diff
@@ -78,6 +85,15 @@ impl<Scalar: PrimeField> MultilinearPolynomial<Scalar> {
       .par_iter_mut()
       .zip(right.par_iter())
       .for_each(|(a, b)| {
+        // The right evals include all terms 'active' in the non-top vars (left evals) plus the corresponding terms of the top var.
+        // To isolate only the relevant terms of the top var, we subtract out `a` to get `b - a`.
+        // Then we scale these top-var terms by r: r * (b - a).
+        // Finally we add back in the low terms.
+        // Put differently, since we only want to bind the top var, we subtract the other terms before scaling it, and add them back afterward.
+        // This gives a + r(b - a) as the general expression.
+        // We iterate over all low terms, adjusting them for the bound top var, whose evals will be removed.
+        // This has the net effect of leaving the sum of the new evals equal to the sum of the old evals when the top var is bound to `r` -- as desired.
+        // [This algorithm is described in the proof of Lemma 4.3 of Thaler's Proofs, Arguments, and Zero-Knowledge.]
```

Comment:
A general remark is that these comments go into a lot of detail, as they assume little prior knowledge from the reader. This has good and bad sides.

One alternative way of making the code more approachable may be to lean on documents where the algorithmics are explained in more detail, as is done here. Should we want to take this approach, Lemma 3.8 of PAZK comes to mind for the memoized multilinear extension, and I have also found this note helpful: http://bgillespie.com/doc/multilinear-evaluation.pdf